### ISSUE TITLE
Traefik fails to add Vary: Accept-Encoding on gzip'd responses

### ISSUE THREAD

**Issue by zarqman** [traefik/traefik#6450]
### Do you want to request a *feature* or report a *bug*?

Bug

### What did you do?

Enabled compression with `compress = true`.

### What did you expect to see?

When Traefik dynamically gzips the response, I expect to see an added `Vary: Accept-Encoding` header (or Accept-Encoding added to an existing Vary header).

### What did you see instead?

Vary is missing or unmodified (if already present with another header).

### Output of `traefik version`: (_What version of Traefik are you using?_)

1.7.20

### What is your environment & configuration (arguments, toml, provider, platform, ...)?

```toml
[entryPoints.https]
  compress = true
```

As long as the client/browser is talking directly to Traefik, all is fine. However, when using a CDN in between, some CDNs (such as CloudFront) cache gzip'd and non-gzip'd content separately, but only if Vary: Accept-Encoding is present. If not present, then the CDN will cache the first request and serve it up to additional clients. If not-gzip'd, this likely just wastes bandwidth. If that first request was gzip'd and a later client is unable to handle a gzip'd response (perhaps a weird API clients, not browser), it will still be sent the gzip'd payload anyway.

**Comment by kepka42** [comment]
Hi! I tried to reproduce this bug and I didn't succeed. 
My config:
1. Started Nginx with static content without gzip compress.
2. Started Traefik with compress = true parameter.

And Traefik returns response with header "Vary: Accept-Encoding". This works as it should.
In other words Traefik must respond in your CDN with Vary header. Or no ...?

**Comment by jbdoumenjou** [comment]
Hi @zarqman,
I made some tests with Traefik v1.7.20 and traefik v2.1.6 and like @kepka42 , I didn't find any issues.
if compress is activated on Traefik, it will add the Accept-Encoding in the Vary Header response.

So, I suspect an issue with your CDN configuration.
Could you explain your use case step by step with the command/tool used to make the requests and show your logs ?

**Comment by zarqman** [comment]
Apologies. I dug in further to find the minimal config necessary to reproduce it and it's a bit more nuanced than I realized.

Summary: when `[retry] attempts = 3` is present _and_ the backend response includes a pre-existing Vary header (eg: `Vary: Origin`), then Traefik 1.7.22 fails to add `Vary: Accept-Encoding`.

###### traefik.toml
```toml
checkNewVersion = false
[entryPoints]
  [entryPoints.http]
    address = ":80"
    compress = true

# this is the problem trigger
[retry]
  attempts = 3

[file]

[backends]
  [backends.a]
    [backends.a.servers.s0]
      url = "http://localhost:4567"

[frontends]
  [frontends.a]
    entryPoints = ["http"]
    backend = "a"
```

###### server.rb - example web server
```ruby
#!/usr/bin/env ruby
require 'sinatra'

get '/' do
  "Hello! #{'a'*1500}\n"
end

get '/vary' do
  headers 'Vary' => 'Origin'
  "Hello! #{'a'*1500}\n"
end
```

Traefik cmdline: `./traefik_darwin-amd64 --logLevel=INFO --accesslog --traefiklog`
Ruby cmdline: `ruby ./server.rb`
  (must run `gem install sinatra webrick` once in advance)

##### Test without backend's own Vary
Does the right thing:
```
$ curl -i --compress http://localhost:80/
HTTP/1.1 200 OK
Content-Encoding: gzip
Content-Type: text/html;charset=utf-8
Date: Thu, 19 Mar 2020 17:06:08 GMT
Server: WEBrick/1.4.2 (Ruby/2.6.5/2019-10-01)
Vary: Accept-Encoding
X-Content-Type-Options: nosniff
X-Frame-Options: SAMEORIGIN
X-Xss-Protection: 1; mode=block
Content-Length: 43

Hello! aa..[snip]..aa
```

##### Test with backend's own Vary
Breaks (but only if [retry] attempts=3 is in traefik.toml):
```
$ curl -i --compress http://localhost:80/vary
HTTP/1.1 200 OK
Content-Encoding: gzip
Content-Type: text/html;charset=utf-8
Date: Thu, 19 Mar 2020 17:06:06 GMT
Server: WEBrick/1.4.2 (Ruby/2.6.5/2019-10-01)
Vary: Origin
X-Content-Type-Options: nosniff
X-Frame-Options: SAMEORIGIN
X-Xss-Protection: 1; mode=block
Content-Length: 43

Hello! aa..[snip]..aa
```

Whether the client actually requests compression (curl's --compress flag) actually doesn't seem to make a difference.

Hope that's everything you need to reproduce, and thanks for looking into it.

**Comment by jbdoumenjou** [comment]
Thank you,
I was able to reproduce. I will investigate further.

**Comment by rtribotte** [comment]
Hello @zarqman,

I give it a try with the latest v2.10 and the issue seems to be solved.
It has most probably been fixed with the changes of https://github.com/traefik/traefik/pull/8245.

Fixed by #8245
